uqcsbot/advent.py

```diff
@@ -183,296 +288,654 @@ def __init__(self, bot: UQCSBot):
-            part_2 = format_seconds(m.day_times.get(2))
-            delta = format_seconds(m.day_delta)
-            return f"{i:>3}) {part_1:>8} {part_2:>8}  {delta:>8}  {m.name}"
+        members = self._get_members(year)
+        if aoc_name not in [member.name for member in members]:
+            await interaction.edit_original_response(
+                content=f"Could not find the Advent of Code name `{aoc_name}` within the UQCS leaderboard."
+            )
+            return
+        member = [member for member in members if member.name == aoc_name]
+        if len(member) != 1:
+            await interaction.edit_original_response(
+                content=f"Could not find a unique Advent of Code name `{aoc_name}` within the UQCS leaderboard."
+            )
+        member = member[0]
+        AOC_id = member.id
+
+        if (
+            query := db_session.query(AOCRegistrations)
+            .filter(
+                and_(
+                    AOCRegistrations.year == year, AOCRegistrations.aoc_userid == AOC_id
+                )
+            )
+            .one_or_none()
+        ) is not None:
+            await interaction.edit_original_response(
+                content=f"Advent of Code name `{aoc_name}` is already registered to <@{query.discord_userid}>. Please contact committee if this is your Advent of Code name."
+            )
+            return
 
-        header = "       Part 1   Part 2     Delta\n"
-        return header + "\n".join(format_member(i, m) for i, m in enumerate(members, 1))
+        discord_id = interaction.user.id
+        if (
+            query := db_session.query(AOCRegistrations)
+            .filter(
+                and_(
+                    AOCRegistrations.year == year,
+                    AOCRegistrations.discord_userid == discord_id,
+                )
+            )
+            .one_or_none()
+        ) is not None:
+            await interaction.edit_original_response(
+                content=f"Your discord account (<@{discord_id}>) is already registered to the Advent of Code name `{query.aoc_userid}`. You'll need to unregister to change name."
+            )
+            return
 
-    def format_advent_leaderboard(
-        self, members: List[Member], is_day: bool, is_global: bool, sort: SortMode
-    ) -> str:
-        """
-        Returns a leaderboard for the given members with the given options.
+        db_session.add(
+            AOCRegistrations(
+                id=id, aoc_userid=AOC_id, year=year, discord_userid=discord_id
+            )
+        )
+        db_session.commit()
+        db_session.close()
+
+        await interaction.edit_original_response(
+            content=f"Advent of Code name `{aoc_name}` is now registered to <@{discord_id}>."
+        )
 
-        If full is True, leaderboard will show progress for all days, otherwise one
-        specific day is shown.
+    @app_commands.default_permissions(manage_guild=True)
```

Comment:
`default_permissions` is newly-discovered to be dangerous: see PR #176.

Use `@app_commands.checks.has_permissions` instead, at least until we get another decorator in-place to enforce that the command is actually used in a particular server.